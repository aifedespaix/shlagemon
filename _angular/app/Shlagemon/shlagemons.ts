import { flanquant, plastoc, ShlagemonType, vapotonie } from "./shlagemons-type";

export interface BaseShlagemon {
    id: string;
    name: string;
    color: string;
    description: string;
    type: ShlagemonType;
}

export const carapouffe: BaseShlagemon = {
    id: 'carapouffe',
    name: 'Carapouffe',
    color: '#333388',
    description: `Carapouffe est une Shlagémone de type Slope. Semi-enfoncée dans sa propre carapace moelleuse, elle ne se déplace qu’en roulant lentement, laissant derrière elle une traînée de paillettes et de gloss fondu. Son maquillage dégouline en permanence, formant une couche protectrice impénétrable — les scientifiques appellent ça le « fard d’armure ».
  Dotée d’un regard mi-séduisant, mi-comateux, elle hypnotise ses adversaires en leur lançant des œillades flasques, accompagnées d’un soupir de lassitude cosmique. Elle passe ses journées à se recoiffer sans bouger la tête, grâce à un système complexe de brosses dissimulées dans son chignon. Sa voix est rauque, son parfum est toxique, et sa principale attaque, "Écrasement Moussant", consiste à s’écrouler violemment sur son ennemi en faisant claquer ses faux ongles.`,
    type: plastoc
};

export const salamiches: BaseShlagemon = {
    id: 'salamiches',
    name: 'Salamiches',
    color: '#ff5533',
    description: `Salamiches est un Shlagémon de type Pain... mais pas le feu noble, non, plutôt le feu de briquet BIC vide qu’on rallume à la 27ᵉ tentative. Toujours torse nu, même en hiver, il arbore une crête en feu mal fixée et des lunettes de soleil qui fondent partiellement à chaque attaque.
  Il se nourrit exclusivement de chips au paprika, de whisky bon marché, et d’approbation sociale. On le trouve souvent en train de se filmer en selfie pendant qu’il crache des étincelles sur des poubelles. Sa capacité spéciale, *Feu d’ego*, augmente sa puissance chaque fois qu’il se fait huer.
  Il envoie des punchlines nulles entre deux jets de flamme molle, et rêve de devenir influenceur muscu alors qu’il n’a que deux abdos (et encore, en ombrage). C’est un être loyal, sauf si on lui propose un kebab gratuit ou une promotion chez Feu Vert.`,
    type: flanquant
};

export const bulgrosboule: BaseShlagemon = {
    id: 'bulgrosboule',
    name: 'Bulgrosboule',
    color: '#88ccff',
    description: `Bulgrosboule est un Shlagémon de type Slope, connu pour ses fesses titanesques capables d’éclipser le soleil couchant. Il avance à reculons, plus par fierté que par stratégie, laissant échapper des bulles parfumées d’une zone que les dresseurs préfèrent ne pas mentionner.

Son cri ressemble à un bain moussant sous pression, et sa capacité signature, *Éruption Fessale*, propulse ses ennemis dans une brume tiède et collante. Doté d’une peau rebondie comme une piscine gonflable de brocante, il adore rebondir sur place en gloussant, ce qui désoriente la plupart des adversaires.

Bulgrosboule est très affectueux, surtout avec ceux qui le massent. Attention toutefois : s’il se met à trembler des miches, c’est trop tard. Il va buller.`,
    type: vapotonie
};
