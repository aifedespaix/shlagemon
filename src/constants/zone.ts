export const VILLAGE_OFFSET = 8
