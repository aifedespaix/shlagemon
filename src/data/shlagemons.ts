import type { BaseShlagemon } from '~/types'
import { shlagemonTypes } from './shlagemons-type'

export const carapouffe: BaseShlagemon = {
  id: 'carapouffe',
  name: 'Carapouffe',
  color: '#333388',
  description: `Carapouffe est une Shlagémon de type Plastoc. Semi-enfoncée dans sa propre carapace moelleuse, elle ne se déplace qu’en roulant lentement, laissant derrière elle une traînée de paillettes et de gloss fondu. Son maquillage dégouline en permanence, formant une couche protectrice impénétrable — les scientifiques appellent ça le « fard d’armure ». Dotée d’un regard mi-séduisant, mi-comateux, elle hypnotise ses adversaires en leur lançant des œillades flasques, accompagnées d’un soupir de lassitude cosmique. Elle passe ses journées à se recoiffer sans bouger la tête, grâce à un système complexe de brosses dissimulées dans son chignon. Sa voix est rauque, son parfum est toxique, et sa principale attaque, "Écrasement Moussant", consiste à s’écrouler violemment sur son ennemi en faisant claquer ses faux ongles.`,
  types: [shlagemonTypes.eau],
  evolution: [],
  coefficient: 1,
}

export const salamiches: BaseShlagemon = {
  id: 'salamiches',
  name: 'Salamiches',
  color: '#ff5533',
  description: `Salamiches est un Shlagémon de type Flanquant... il brûle, mais pas de feu noble, non, plutôt de feu de briquet BIC vide qu’on rallume à la 27ᵉ tentative. Toujours torse nu, même en hiver, il arbore une crête en feu mal fixée et des lunettes de soleil qui fondent partiellement à chaque attaque. Il se nourrit exclusivement de chips au paprika, de whisky bon marché, et d’approbation sociale. On le trouve souvent en train de se filmer en selfie pendant qu’il crache des étincelles sur des poubelles. Sa capacité spéciale, *Feu d’ego*, augmente sa puissance chaque fois qu’il se fait huer. Il envoie des punchlines nulles entre deux jets de flamme molle, et rêve de devenir influenceur muscu alors qu’il n’a que deux abdos (et encore, en ombrage). C’est un être loyal, sauf si on lui propose un kebab gratuit ou une promotion chez Feu Vert.`,
  types: [shlagemonTypes.feu],
  coefficient: 1,
}

export const bulgrosboule: BaseShlagemon = {
  id: 'bulgrosboule',
  name: 'Bulgrosboule',
  color: '#88ccff',
  description: `Bulgrosboule est un Shlagémon de type Vapotonie, connu pour ses fesses titanesques capables d’éclipser le soleil couchant. Il avance à reculons, plus par fierté que par stratégie, laissant échapper des bulles parfumées d’une zone que les dresseurs préfèrent ne pas mentionner. Son cri ressemble à un bain moussant sous pression, et sa capacité signature, *Éruption Fessale*, propulse ses ennemis dans une brume tiède et collante. Doté d’une peau rebondie comme une piscine gonflable de brocante, il adore rebondir sur place en gloussant, ce qui désoriente la plupart des adversaires. Bulgrosboule est très affectueux, surtout avec ceux qui le massent. Attention toutefois : s’il se met à trembler des miches, c’est trop tard. Il va buller.`,
  types: [shlagemonTypes.plante],
  coefficient: 1,
}

export const pikachiant: BaseShlagemon = {
  id: 'pikachiant',
  name: 'Pikachiant',
  color: '#f9e743',
  description: `Pikachiant est un Shlagémon de type Electrochiasse, capable de te saouler avant même de lancer une attaque. Sa queue ressemble à une antenne de TNT tordue, et il émet des bips stridents dès qu’il est contrarié. Il se décharge en permanence (surtout en présence d’appareils électroniques de valeur), et son attaque signature, *Charge Sociale*, provoque une gêne électrique dans tout un rayon de 3 mètres.
Il vit généralement dans des squats connectés où il recharge ses batteries avec des câbles USB volés dans les trains. Il pense que TikTok est une source d’énergie et parle en onomatopées gênantes. À ne pas confondre avec Pikachu, même si lui aussi finit souvent dans une Poképrison pour tapage nocturne.`,
  types: [shlagemonTypes.electrique],
  coefficient: 2,
}

export const mewteub: BaseShlagemon = {
  id: 'mewteub',
  name: 'Mewteub',
  color: '#f9e743',
  description: `Mewteub est un Shlagémon de type Slope, né d’un bug génétique entre une prise HDMI et une entité cosmique sous anxiolytiques. On murmure qu’il aurait été cloné à partir d’un vieux câble péritel tombé dans une bassine de Monster Energy. Il flotte constamment, mais sans élégance — plutôt comme une chaussette dans un jacuzzi. Son corps translucide laisse entrevoir un réseau veineux qui clignote au rythme d’un modem 56k. Il émet des sons télépathiques, tous plus gênants les uns que les autres : soupirs, "hmm" douteux et parfois des extraits de discussions MSN de 2003. Il communique principalement par vibration de teub psychique, ce qui perturbe fortement les autres Shlagémons — et les enfants non accompagnés. Sa capacité spéciale, *Onde Malaise*, fige l’adversaire dans une gêne abyssale pendant plusieurs tours. On dit que croiser son regard provoque des flashbacks de soirées gênantes en boîte de province.
Mewteub est rare. Trop rare. Et franchement, c’est peut-être mieux ainsi.`,
  types: [shlagemonTypes.psy],
  coefficient: 10,
}

export const sacdepates: BaseShlagemon = {
  id: 'sacdepates',
  name: 'Sac de Pâtes',
  color: '#f9e743',
  description: `Sac de Pâtes est une boule vivante de spaghettis emmêlés, dont les longs brins forment un labyrinthe mouvant. Doté de deux yeux perçants incrustés au milieu de ses pâtes, il intimide quiconque croise son regard infernal. Ses pieds rouges, lisses et luisants, lui permettent de rouler à toute vitesse sur ses adversaires, qu’il écrase sans pitié dans un rire aigu et diabolique. Il passe ses journées à se peigner minutieusement avec un peigne fin, espérant un jour démêler le nœud infini qu’il est devenu. On raconte que plus ses spaghettis sont emmêlés, plus il devient redoutable. Talent : Nœud Fatal — Quand Sacdepâtes subit une attaque physique, il peut s’enrouler autour de l’ennemi pour le piéger et l’immobiliser.`,
  types: [shlagemonTypes.plante],
  coefficient: 2,
}

export const rouxPasCool: BaseShlagemon = {
  id: 'roux-pas-cool',
  name: 'Roux pas Cool',
  color: '#f9e743',
  description: `Roux pas Cool est un Pokémon Balladépressif de type acoustique / malaise. Anciennement cool, Roux pas Cool a passé trop de temps à gratter des accords mineurs au bord d’un volcan éteint. Désormais, Roux pas Cool erre avec une guitare trop grande pour ses ailes, des taches de rousseur qui pleurent, et une chemise à carreaux qui sent l’herbe humide et les regrets. Son plumage a pris une teinte rouille triste, et sa mèche rousse cache un regard empli de remords, comme s’il réalisait constamment qu’il aurait pu évoluer en rapace légendaire, mais a préféré sortir un EP en indépendant. Il fait toujours un peu froid autour de lui, même en plein été. Sa capacité signature, Refrain Gênant, inflige un malaise profond à toute l’arène, réduisant la précision des attaques ennemies tant qu’ils détournent le regard. Il est très doué pour faire fuir les Pokémon sauvages… et les rendez-vous galants.`,
  types: [shlagemonTypes.vol],
  coefficient: 2,
}

export const canarchichon: BaseShlagemon = {
  id: 'canarchicon',
  name: 'Canarchicon',
  color: '#f9e743',
  description: `Canarchicon est un Shlagémon de type Cuisine, cousin dégénéré du célèbre canard qu'on ne nommera pas pour des raisons juridiques. Toujours armé de son poireau fatigué (qu’il appelle tendrement “Jean-Chibre”), il chancelle d’un pas bancal, probablement à cause de ses soirées passées à picoler du bouillon cube fermenté. Son œil au beurre noir laisse deviner un mode de vie instable, fait de bastons derrière des bennes à frites et de paris perdus contre des Roucool. Il ne vole pas, il flotte à moitié — sans réelle direction — au gré des vents et des vapeurs d’alcool de cuisson. Sa capacité signature, *Coup de Poireau Tournoyant*, inflige peu de dégâts mais une honte durable. Il peut aussi utiliser *Flatulence de Gras*, une attaque à effet de zone olfactif. Canarchichon n’a jamais gagné un seul combat, mais il persiste... parce qu’il a oublié qu’il pouvait abandonner.`,
  types: [shlagemonTypes.normal, shlagemonTypes.vol],
  coefficient: 2,
}

export const nosferachid: BaseShlagemon = {
  id: 'nosferachid',
  name: 'Nosferachid',
  color: '#f9e743',
  description: `Nosferachid est un Shlagémon de type Slope, nocturne et fier de l’être. Originaire d’un quartier poussiéreux à l’ombre des HLM de Kanto-Ouest, il vole en rasant les murs avec sa sacoche Louis Viton™ (orthographe non contractuelle), achetée 7 Pokédollars au marché aux puces de Celadop-les-Bains. Il arbore fièrement une casquette visière plate LV assortie, portée en biais pour maximiser l’aérodynamisme et l’attitude. Derrière ses grandes ailes décharnées se cache un petit sachet de "poudre mystique", qu’il dit être pour ses Pokémon... mais personne n’y croit. Nosferachid passe ses nuits à zoner en cercle autour des lampadaires, tout en freestyleant des menaces inaudibles à base de “j’te croque ta daronne” en ultrason. Il attaque rarement de front : il préfère t’endormir avec *Haleine Mentholée Menthe Forte* avant de te vider ton inventaire. On dit qu’il a déjà mordu un Ronflex pour une histoire de briquet. Ronflex dort encore, mais pas pour les bonnes raisons.`,
  types: [shlagemonTypes.poison, shlagemonTypes.vol],
  coefficient: 2,
}

export const dartagnan: BaseShlagemon = {
  id: 'dartagnan',
  name: 'D\'Art Tagnan',
  color: '#f9e743',
  description: `D\'Art Tagnan est un Mousquépique de type panache combatif. Toujours prêt à pérorer avant de piquer, il surgit d’un nuage de poussière dramatique en criant « En garde, manant ! », alors que personne ne l’a regardé. Avec ses dards en forme de rapières et ses antennes sculptées en bouclettes, il enchaîne les moulinets dans le vide juste pour le style. Sa moustache n’existe pas, mais il la twiste régulièrement du bout des griffes, persuadé que ça le rend irrésistible. Son chapeau à plume est greffé directement sur son crâne depuis sa naissance — selon la légende, il serait sorti de son cocon en criant « À l’attaque pour l’honneur et les gaufres ! » Il défie les Pokémon sauvages à des duels de poésie, vole au secours des baies tombées au sol, et fond en larmes si on lui abîme sa cape. Sa technique signature, Tournoyement Galant, consiste à piquer son adversaire après avoir tourné sur lui-même au moins huit fois, tout en citant du théâtre. Son flair pour le drame est tel que certains chercheurs pensent qu’il est en fait mi-insecte, mi-acteur raté.`,
  types: [shlagemonTypes.poison],
  coefficient: 2,
}

export const ptitocard: BaseShlagemon = {
  id: 'ptitocard',
  name: 'Ptitocard',
  color: '#f9e743',
  description: `Ptitocard est un Shlagémon de type Slope, aussi fragile qu’une biscotte mouillée et aussi expressif qu’un poisson-panique en pleine crise existentielle. Son regard vide, humide et terriblement plaintif fait fondre les cœurs les plus endurcis — ou les agace profondément, au choix. Il coule plus qu’il ne nage, et sa spirale ventrale ne tourne que lorsqu’il fait une crise d’angoisse. Il bave en permanence, mais pas de la bouche : c’est tout son corps qui transpire la détresse. On pense qu’il est triste de naissance, mais certains spécialistes évoquent une simple allergie à la vie. Sa capacité spéciale, *Larme Infinie*, provoque l’ennui mortel chez l’ennemi. Un adversaire qui regarde Ptitocard pendant plus de 10 secondes peut tomber dans un coma d’indifférence profonde. Ptitocard rêve de devenir un grand champion... mais ne fait rien pour. On le trouve souvent flottant à la surface des flaques, en train de se demander s’il mérite vraiment d’évoluer. Spoiler : pas sûr.`,
  types: [shlagemonTypes.eau],
  coefficient: 2,
}

export const goubite: BaseShlagemon = {
  id: 'goubite',
  name: 'Goubite',
  color: '#f9e743',
  description: `Goubite est un Shlagémon de type Slope, difficile à regarder... et encore plus difficile à oublier. Doté d’un corps long et rosâtre, légèrement bombé sur le haut, il est souvent victime de malentendus anatomiques dans les écoles de dresseurs. Pourtant, lui, il s’assume. Fier, droit, et toujours un peu moite. Sa tête est cerclée d’une barbe flasque, dégoulinante de sueur tiède, qu’il secoue parfois pour marquer son territoire. Il sourit en permanence, d’un air satisfait qui met tout le monde mal à l’aise, même les aveugles. Ses pieds sont fripés, velus, et sentent la crevette oubliée dans une chaussette humide. Il se déplace lentement en produisant un petit bruit mou et répétitif, comparable à un bisou dont on ne veut pas. Sa capacité signature, *Giclée Visqueuse*, inflige des dégâts modérés mais impose un malus mental permanent à ceux qui en sont témoins. Certains affirment qu’il peut aussi lancer *Barbe Emotive*, une attaque qui colle et qui sent la salle de muscu humide. À ne pas confondre avec un objet sexuel. Sauf si vraiment on est pressé de consulter.`,
  types: [shlagemonTypes.feu],
  coefficient: 2,
}

export const alaouakbar: BaseShlagemon = {
  id: 'alaouakbar',
  name: 'Alaouakbar',
  color: '#c5a641',
  description: `Considéré comme un sage dans son quartier... du moins par lui-même. Il prétend maîtriser les arts mystiques du *Contrôle du Destin* mais confond souvent télékinésie et procrastination. On le reconnaît à ses deux grosses cuillères de cantine, tordues à force de remuer du thé aux herbes chelou dans des bouteilles de soda. Vêtu d’une djellaba délavée et d’un regard qui voit à travers toi (mais pas très loin), il marmonne des incantations approximatives pendant qu’il fait tourner sa sacoche à bandoulière en plastique doré. Il prétend méditer, mais il dort à 80%. Sa capacité spéciale, *Chakra Perimé*, déséquilibre l’ennemi avec une odeur de patchouli acide et une attaque psychique floue. Il peut aussi invoquer son attaque signature, *Projection Spirituelle*, qui consiste à lancer une cuillère sur son adversaire en hurlant “vision sacrée !” sans grand effet. On le croise souvent assis sur un banc, seul, en train de discuter avec ses propres Pokéball vides.`,
  types: [shlagemonTypes.psy],
  coefficient: 2,
}

export const starters: BaseShlagemon[] = [carapouffe, salamiches, bulgrosboule]
export const allShlagemons: BaseShlagemon[] = [
  carapouffe,
  salamiches,
  bulgrosboule,
  pikachiant,
  mewteub,
  nosferachid,
  rouxPasCool,
  dartagnan,
  ptitocard,
  sacdepates,
  canarchichon,
  goubite,
  alaouakbar,
]
