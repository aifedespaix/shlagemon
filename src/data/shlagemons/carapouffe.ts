import type { BaseShlagemon } from '~/type'
import { shlagemonTypes } from '../shlagemons-type'
import { carabifle } from './evolutions/carabifle'

export const carapouffe: BaseShlagemon = {
  id: 'carapouffe',
  name: 'data.shlagemons.carapouffe.name',
  description: 'data.shlagemons.carapouffe.description',
  types: [shlagemonTypes.eau],
  evolution: { base: carabifle, condition: { type: 'lvl', value: 16 } },
  speciality: 'evolution0',
}
export default carapouffe
