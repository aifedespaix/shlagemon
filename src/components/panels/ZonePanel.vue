<script setup lang="ts">
import type { Zone } from '~/type'
import { computed } from 'vue'
import Button from '~/components/ui/Button.vue'
import { useMainPanelStore } from '~/stores/mainPanel'
import { useShlagedexStore } from '~/stores/shlagedex'
import { useTrainerBattleStore } from '~/stores/trainerBattle'
import { useZoneStore } from '~/stores/zone'
import { useZoneProgressStore } from '~/stores/zoneProgress'

const zone = useZoneStore()
const dex = useShlagedexStore()
const panel = useMainPanelStore()
const progress = useZoneProgressStore()
const trainerBattle = useTrainerBattleStore()

const xpZones = computed(() => zone.zones.filter(z => z.maxLevel > 0))

function canAccess(z: Zone) {
  if (z.type === 'village')
    return z.minLevel <= dex.highestLevel
  const idx = xpZones.value.findIndex(x => x.id === z.id)
  if (idx === 0)
    return dex.highestLevel >= z.minLevel
  const prev = xpZones.value[idx - 1]
  return dex.highestLevel >= z.minLevel && progress.isKingDefeated(prev.id)
}

function onAction(id: string) {
  if (id === 'shop')
    panel.showShop()
  else if (id === 'explore')
    panel.showTrainerBattle()
}

function fightKing() {
  const trainer = zone.getKing(zone.current.id)
  trainerBattle.setQueue([trainer])
  panel.showTrainerBattle()
}

function classes(z: Zone) {
  const classes = []
  z.id === zone.current.id ? classes.push('bg-primary text-dark dark:bg-light') : classes.push('bg-gray-200 dark:bg-gray-700')
  if (z.type === 'village') {
    classes.push('bg-green-300 dark:bg-green-800')
  }
  return classes.join(' ')
}
</script>

<template>
  <div class="flex flex-col gap-2" md="gap-3">
    <div class="flex flex-wrap justify-center gap-1" md="gap-2">
      <button
        v-for="z in zone.zones"
        :key="z.id"
        class="rounded px-2 py-1 text-xs"
        :class="[classes(z), { 'opacity-50 cursor-not-allowed': !canAccess(z) }]"
        :disabled="!canAccess(z)"
        @click="zone.setZone(z.id)"
      >
        {{ z.name }}
      </button>
    </div>
    <div class="flex flex-col items-center gap-1" md="gap-2">
      <Button
        v-for="action in zone.current.actions"
        :key="action.id"
        class="text-xs"
        @click="onAction(action.id)"
      >
        {{ action.label }}
      </Button>
      <Button
        v-if="!progress.isKingDefeated(zone.current.id) && progress.canFightKing(zone.current.id)"
        class="text-xs"
        @click="fightKing"
      >
        Vaincre le roi de la zone
      </Button>
      <div v-else-if="progress.isKingDefeated(zone.current.id)" class="text-xs font-bold">
        Roi vaincu !
      </div>
    </div>
  </div>
</template>
