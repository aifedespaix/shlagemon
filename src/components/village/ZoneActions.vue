<script setup lang="ts">
import { useArenaStore } from '~/stores/arena'
import { useDialogStore } from '~/stores/dialog'
import { useMainPanelStore } from '~/stores/mainPanel'
import { useMiniGameStore } from '~/stores/miniGame'
import { useTrainerBattleStore } from '~/stores/trainerBattle'
import { useZoneStore } from '~/stores/zone'
import { useZoneProgressStore } from '~/stores/zoneProgress'

const zone = useZoneStore()
const panel = useMainPanelStore()
const mini = useMiniGameStore()
const progress = useZoneProgressStore()
const trainerBattle = useTrainerBattleStore()
const arena = useArenaStore()
const dialog = useDialogStore()

const hasKing = computed(() =>
  zone.current.hasKing ?? zone.current.type === 'sauvage',
)
const hasArena = computed(() => !!zone.current.arena)
const hasPoulailler = computed(() => !!zone.current.village?.poulailler)
const arenaCompleted = computed(() => progress.isArenaCompleted(zone.current.id))
const currentKing = computed(() =>
  hasKing.value ? zone.getKing(zone.current.id) : undefined,
)
const kingLabel = computed(() =>
  currentKing.value?.character.gender === 'female' ? 'reine' : 'roi',
)

function actionIcon(id: string) {
  switch (id) {
    case 'shop':
      return 'i-carbon:shopping-bag'
    case 'explore':
      return 'i-mdi:compass'
    case 'minigame':
      return 'i-carbon:game-console'
    default:
      return ''
  }
}

function onAction(id: string) {
  if (arena.inBattle)
    return
  if (id === 'shop') {
    panel.showShop()
  }
  else if (id === 'explore') {
    panel.showTrainerBattle()
  }
  else if (id === 'minigame') {
    if (zone.current.miniGame)
      mini.select(zone.current.miniGame)
    panel.showMiniGame()
  }
}

function openArena() {
  if (arena.inBattle)
    return
  const data = zone.current.arena?.arena
  if (data)
    arena.setArena(data)
  dialog.resetArenaDialogs()
  panel.showArena()
}

function fightKing() {
  if (arena.inBattle)
    return
  const trainer = currentKing.value
  if (!trainer)
    return
  trainerBattle.setQueue([trainer])
  panel.showTrainerBattle()
}

function openPoulailler() {
  if (arena.inBattle)
    return
  panel.showPoulailler()
}
</script>

<template>
  <div class="actions-grid grid w-full gap-2 p-1" md="gap-3 p-2">
    <UiNavigationButton
      v-if="zone.current.village?.shop"
      icon="i-carbon:shopping-bag"
      label="Magasin"
      class="bg-green-600 text-white dark:bg-green-700"
      hover="bg-green-700 dark:bg-green-800"
      :disabled="arena.inBattle"
      @click="panel.showShop()"
    />
    <UiNavigationButton
      v-for="action in zone.current.actions"
      :key="action.id"
      :icon="actionIcon(action.id)"
      :label="action.label"
      :class="action.id === 'minigame' ? 'bg-violet-600 text-white dark:bg-violet-700' : ''"
      :hover="action.id === 'minigame' ? 'bg-violet-700 dark:bg-violet-800' : undefined"
      :disabled="arena.inBattle"
      @click="onAction(action.id)"
    />
    <UiNavigationButton
      v-if="hasArena && !arenaCompleted"
      icon="i-mdi:sword-cross"
      label="Arène"
      class="bg-red-600 text-white dark:bg-red-700"
      hover="bg-red-700 dark:bg-red-800"
      :disabled="arena.inBattle"
      @click="openArena"
    />
    <UiNavigationButton
      v-if="hasPoulailler"
      icon="i-game-icons:bird-house"
      label="Poulailler"
      :disabled="arena.inBattle"
      @click="openPoulailler"
    />
    <UiNavigationButton
      v-if="hasKing && !progress.isKingDefeated(zone.current.id) && progress.canFightKing(zone.current.id)"
      icon="i-game-icons:crown"
      :label="`Défier la ${kingLabel} de la zone`"
      :disabled="arena.inBattle"
      @click="fightKing"
    />
    <div
      v-else-if="hasKing && progress.isKingDefeated(zone.current.id)"
      class="flex-center text-xs font-bold"
    >
      {{ kingLabel.charAt(0).toUpperCase() + kingLabel.slice(1) }}
      vaincu{{ kingLabel === 'reine' ? 'e' : '' }} !
    </div>
  </div>
</template>

<style scoped>
.actions-grid {
  grid-template-columns: repeat(auto-fill, minmax(100px, 1fr));
}
</style>
