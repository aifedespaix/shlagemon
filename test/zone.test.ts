import { mount } from '@vue/test-utils'
import { createPinia, setActivePinia } from 'pinia'
import { describe, expect, it, vi } from 'vitest'
import ZonePanel from '../src/components/panel/Zone.vue'
import ZoneActions from '../src/components/village/ZoneActions.vue'
import { carapouffe } from '../src/data/shlagemons/carapouffe'
import { useMainPanelStore } from '../src/stores/mainPanel'
import { useShlagedexStore } from '../src/stores/shlagedex'
import { useZoneStore } from '../src/stores/zone'
import { useZoneProgressStore } from '../src/stores/zoneProgress'
import { xpForLevel } from '../src/utils/dexFactory'

describe('zone store', () => {
  it('changes current zone', () => {
    setActivePinia(createPinia())
    const store = useZoneStore()
    expect(store.current.id).toBe(store.zones[0].id)
    store.setZone('chemin-du-slip')
    expect(store.current.id).toBe('chemin-du-slip')
  })
})

describe('zone panel', () => {
  it('renders actions', () => {
    const pinia = createPinia()
    setActivePinia(pinia)
    const wrapper = mount(ZonePanel, {
      global: { plugins: [pinia], provide: { selectZone: vi.fn() } },
    })
    // first zone is now the Plaine Kékette
    expect(wrapper.text()).toContain('Plaine Kékette')
  })

  it('filters zones by level', async () => {
    const pinia = createPinia()
    setActivePinia(pinia)
    const dex = useShlagedexStore()
    const progress = useZoneProgressStore()
    const mon = dex.createShlagemon(carapouffe)
    const wrapper = mount(ZonePanel, {
      global: { plugins: [pinia], provide: { selectZone: vi.fn() } },
    })
    let btn = [
      ...wrapper.findAll('#savages button'),
      ...wrapper.findAll('#villages button'),
    ].find(b => b.text().includes('Chemin du Slip'))
    expect(btn).toBeUndefined()
    for (let i = 0; i < 20; i++)
      progress.addWin('plaine-kekette')
    for (let i = 0; i < 20; i++)
      progress.addWin('bois-de-bouffon')
    progress.defeatKing('plaine-kekette')
    progress.defeatKing('bois-de-bouffon')
    for (let i = 0; i < 9; i++)
      await dex.gainXp(mon, xpForLevel(mon.lvl))
    await wrapper.vm.$nextTick()
    btn = [
      ...wrapper.findAll('#savages button'),
      ...wrapper.findAll('#villages button'),
    ].find(b => b.text().includes('Chemin du Slip'))
    expect(btn).toBeDefined()
  })

  it('shows king button after 20 wins', async () => {
    const pinia = createPinia()
    setActivePinia(pinia)
    const progress = useZoneProgressStore()
    const zone = useZoneStore()
    const wrapper = mount(ZoneActions, { global: { plugins: [pinia] } })
    zone.setZone('plaine-kekette')
    await wrapper.vm.$nextTick()
    expect(wrapper.text()).not.toContain('Défier le roi')
    for (let i = 0; i < 20; i++)
      progress.addWin('plaine-kekette')
    await wrapper.vm.$nextTick()
    expect(wrapper.text()).toContain('Défier la roi de la zone')
  })

  it('disables zone buttons during trainer battle', async () => {
    const pinia = createPinia()
    setActivePinia(pinia)
    const dex = useShlagedexStore()
    dex.createShlagemon(carapouffe)
    const panel = useMainPanelStore()
    const wrapper = mount(ZonePanel, {
      global: { plugins: [pinia], provide: { selectZone: vi.fn() } },
    })
    panel.showTrainerBattle()
    await wrapper.vm.$nextTick()
    const buttons = [
      ...wrapper.findAll('#savages button'),
      ...wrapper.findAll('#villages button'),
    ]
    expect(buttons.length).toBeGreaterThan(0)
    expect(buttons.every(b => b.attributes('disabled') !== undefined)).toBe(true)
  })

  it('disables zone buttons when dialog is visible', async () => {
    const pinia = createPinia()
    setActivePinia(pinia)
    const wrapper = mount(ZonePanel, {
      global: { plugins: [pinia], provide: { selectZone: vi.fn() } },
    })
    await wrapper.vm.$nextTick()
    const buttons = [
      ...wrapper.findAll('#savages button'),
      ...wrapper.findAll('#villages button'),
    ]
    expect(buttons.length).toBeGreaterThan(0)
    expect(buttons.every(b => b.attributes('disabled') !== undefined)).toBe(true)
  })
})
