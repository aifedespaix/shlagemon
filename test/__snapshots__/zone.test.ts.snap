// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`zone panel > renders actions 1`] = `
"<div class="flex flex-col gap-2" md="gap-3">
  <div class="flex flex-wrap justify-center gap-1" md="gap-2"><button class="rounded px-2 py-1 text-xs bg-primary text-white">Village Paumé</button><button class="rounded px-2 py-1 text-xs bg-gray-200 dark:bg-gray-700">Grotte Sombre</button><button class="rounded px-2 py-1 text-xs bg-gray-200 dark:bg-gray-700">Plaine Verdoyante</button></div>
  <div class="flex flex-col items-center gap-1" md="gap-2"><span class="font-bold">Village Paumé</span><button class="rounded px-2 py-1 text-white bg-cyan-600 dark:bg-cyan-700 hover:bg-cyan-700 dark:hover:bg-cyan-800/80 text-xs">Entrer le Shop</button></div>
</div>
<dialog data-v-b759700c="" class="modal">
  <div data-v-b759700c="" class="modal-content relative"><button data-v-b759700c="" type="button" class="absolute right-2 top-2 h-6 w-6 flex items-center justify-center rounded bg-gray-200 dark:bg-gray-700 hover:bg-gray-300 dark:hover:bg-gray-600"> × </button>
    <div class="h-full flex flex-col">
      <h2 class="mb-2 text-center font-bold"> Boutique </h2>
      <div class="flex flex-col gap-2 overflow-auto">
        <div class="flex items-center gap-2 border rounded bg-white p-2 dark:bg-gray-900"><img src="/items/shlageball/shlageball.png" alt="Shlage Ball" class="h-8 w-8 object-contain">
          <div class="flex flex-1 flex-col text-left"><span class="font-bold">Shlage Ball</span><span class="text-xs">Permet de capturer des Shlagémons sauvages.</span></div><span class="font-bold">10$</span><button class="rounded px-2 py-1 text-white bg-cyan-600 dark:bg-cyan-700 hover:bg-cyan-700 dark:hover:bg-cyan-800/80 ml-2"> Acheter </button>
        </div>
        <div class="flex items-center gap-2 border rounded bg-white p-2 dark:bg-gray-900">
          <!--v-if-->
          <div class="flex flex-1 flex-col text-left"><span class="font-bold">Potion Dégueulasse</span><span class="text-xs">Soigne 50 HP de votre Shlagémon.</span></div><span class="font-bold">5$</span><button class="rounded px-2 py-1 text-white bg-cyan-600 dark:bg-cyan-700 hover:bg-cyan-700 dark:hover:bg-cyan-800/80 ml-2"> Acheter </button>
        </div>
        <div class="flex items-center gap-2 border rounded bg-white p-2 dark:bg-gray-900">
          <!--v-if-->
          <div class="flex flex-1 flex-col text-left"><span class="font-bold">Spray Anti-Odeur</span><span class="text-xs">Augmente temporairement la défense.</span></div><span class="font-bold">7$</span><button class="rounded px-2 py-1 text-white bg-cyan-600 dark:bg-cyan-700 hover:bg-cyan-700 dark:hover:bg-cyan-800/80 ml-2"> Acheter </button>
        </div>
      </div>
    </div>
  </div>
</dialog>"
`;
